### PR TITLE
improve clone impl as per https://github.com/Lokathor/tinyvec/issues/143

### PR DESCRIPTION
**Pull request by Lokathor**
* Closes https://github.com/Lokathor/tinyvec/issues/143

### CODE REVIEW

**Comment by Shnatsel** [comment]
Wait, isn't `clone_from` supposed to discard the data that was previously contained by the collection? Why is there `zip` then?

**Comment by Shnatsel** [comment]
I wonder if a fast path for `Copy` types is necessary, for cases like a long-ish vec of `u64`? I'm not sure the current implementation would lower to a `memcpy()`

**Comment by Lokathor** [comment]
Best as I understand, this is improving the situation in some but not all situations. If this code is bad for a particular copy type, then it was *previously* also bad for that type.

File: src/arrayvec.rs [hunk 1]
```diff
@@ -101,12 +101,43 @@ macro_rules! array_vec {
...
+    let iter = self
+      .data
+      .as_slice_mut()
+      .iter_mut()
+      .zip(o.data.as_slice())
+      .take(self.len.max(o.len) as usize);
+    for (dst, src) in iter {
+      dst.clone_from(src)
+    }
+    self.len = o.len;
```
Comment (Nemo157) [inline src/arrayvec.rs:130]: Might want to take-and-drop all the elements from `o.len..self.len`.

Comment (Soveu) [reply src/arrayvec.rs:130]: Just remember to skip this part if `o.len >= self.len`

Comment (Lokathor) [reply src/arrayvec.rs:130]: If I follow you right, we'd want to have something like:

if o.len < self.len {
  self[o.len..].iter_mut().for_each(|r| take(r));
}

immediately after the for loop?

Comment (Soveu) [reply src/arrayvec.rs:130]:
```rust
if let Some(to_drop) = self.data.get_mut(o.len() .. self.len()) {
  to_drop.for_each(|x| take(x));
}
```

**Pull request merged**